### PR TITLE
Submissions persistence

### PR DESCRIPTION
**Pull request by Sagolbah**
Persist information about submissions in database.
To be implemented: Line of user's submissions

### CODE REVIEW

**Comment by Sagolbah** [comment]
Warning: it's really huge, and RunningVerdict logic is changed: when it's not persisted, it holds a placeholder empty list.

File: html-server/src/main/kotlin/ru/testing/html/controller/Routing.kt [hunk 1]
```diff
@@ -188,7 +193,7 @@ private suspend fun PipelineContext<Unit, ApplicationCall>.receiveTask(configura
             require(task != null) { "Task is not specified" }
             require(source != null) { "Source is not specified" }
             require(language != null) { "Language is not specified" }
-            val submission = makeSubmission(configuration, title, source, fileType = language, task)
+            val submission = makeSubmission(configuration, title, source, fileType = language, task, call.principal<UserPrincipal>()!!.user.id)
```
Comment (vladrus13) [inline html-server/src/main/kotlin/ru/testing/html/controller/Routing.kt]: Need if (call.pricipal<UserPricipal>() == null)?

Comment (zhelenskiy) [reply html-server/src/main/kotlin/ru/testing/html/controller/Routing.kt]: The same question

File: model/src/main/kotlin/ru/testing/databese/DatabaseInitializer.kt [hunk 2]
```diff
@@ -26,7 +28,7 @@ class DatabaseInitializer : AbstractDatabaseInitializer {
             password = properties.getProperty("password")
         )
         transaction {
-            SchemaUtils.create(Users)
+            SchemaUtils.create(Users, Submissions, TestVerdicts)
         }
     }
 }
```
Comment (zhelenskiy) [inline model/src/main/kotlin/ru/testing/databese/DatabaseInitializer.kt]: Add an empty line at the EOF

File: model/src/main/kotlin/ru/testing/databese/ResultHolder.kt [hunk 3]
```diff
@@ -77,10 +91,48 @@ class ResultHolder : AbstractResultHolder {
...
+    }
+
+    private fun doSendVerdict(submissionId: Long, verdict: SubmissionVerdict) {
+        transaction {
+            Submissions.update({ Submissions.id eq submissionId }) {
+                it[Submissions.serializedStatus] = Json.encodeToString(verdict)
+            }
         }
     }
 }
```
Comment (zhelenskiy) [inline model/src/main/kotlin/ru/testing/databese/ResultHolder.kt]: Add an empty line at the end

File: model/src/main/kotlin/ru/testing/databese/UserHolder.kt [hunk 4]
```diff
@@ -23,7 +23,7 @@ class UserHolder : AbstractUserHolder {
 
     override fun createUser(name: String, passwordHash: String) {
         if (findUserByName(name) != null) {
-            throw RuntimeException("This username is occupied")  // TODO: probably, can be better, don't know Kotlin specifics :(
```
Comment (zhelenskiy) [inline model/src/main/kotlin/ru/testing/databese/UserHolder.kt:26]: What is the actual problem here?

File: model/src/main/kotlin/ru/testing/databese/definition/Submissions.kt [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+import org.jetbrains.exposed.sql.Table
+
+object Submissions : Table() {
+    val id = long("id").autoIncrement()
+    val userId = long("userId").references(Users.id, onDelete = ReferenceOption.CASCADE)
+    val taskName = varchar("taskName", length = 50)
+    val testCount = integer("testCount")  // Allows partial testing, for example, pretests.
+    val serializedStatus = varchar("serializedStatus", length = 4096);
+    override val primaryKey = PrimaryKey(id)
+}
```
Comment (zhelenskiy) [inline model/src/main/kotlin/ru/testing/databese/definition/Submissions.kt:13]: Empty line

File: model/src/main/kotlin/ru/testing/databese/definition/TestVerdicts.kt [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+import org.jetbrains.exposed.sql.ReferenceOption
+import org.jetbrains.exposed.sql.Table
+
+object TestVerdicts : Table() {
+    val testId = integer("testId")
+    val submissionId = long("submissionId").references(Submissions.id, onDelete = ReferenceOption.CASCADE)
+    val serializedStatus = varchar("serializedStatus", length = 4096)
+    override val primaryKey = PrimaryKey(testId, submissionId)
+}
```
Comment (zhelenskiy) [inline model/src/main/kotlin/ru/testing/databese/definition/TestVerdicts.kt]: Empty line

File: model/src/main/resources/database.properties [hunk 7]
```diff
@@ -1,4 +1,4 @@
 url=jdbc:postgresql://localhost:5432/testsys
 driver=org.postgresql.Driver
-user=vlad
-password=vlad
\ No newline at end of file
+user=sagolbah
+password=
```
Comment (zhelenskiy) [inline model/src/main/resources/database.properties:4]: Perhaps, we can all set properties to:
```
url=jdbc:postgresql://localhost:5432/testsys
driver=org.postgresql.Driver
user=testsys_admin
password=
```
@vladrus13 @Sagolbah

File: polygon/src/main/kotlin/ru/testing/polygon/submission/SubmissionsFactory.kt [hunk 8]
```diff
@@ -19,8 +19,9 @@ fun makeSubmission(
...
     source: String,
     fileType: AbstractTypeOfLaunching,
-    task: Task
+    task: Task,
+    userId: Long
 ): SubmissionFile = with(configuration) {
-    val id = resultHolder.addTask(task)
+    val id = resultHolder.addSubmission(task, userId)
     return SubmissionFile(id, title, source, fileType, task)
 }
```
Comment (zhelenskiy) [inline polygon/src/main/kotlin/ru/testing/polygon/submission/SubmissionsFactory.kt]: Empty line

File: testlib/build.gradle.kts [hunk 9]
```diff
@@ -12,5 +13,6 @@ repositories {
 }
 
 dependencies {
+    implementation("org.jetbrains.kotlinx:kotlinx-serialization-json:1.3.1")
     implementation("io.ktor:ktor-html-builder:1.6.4")
 }
```
Comment (zhelenskiy) [inline testlib/build.gradle.kts]: Empty line

File: testlib/src/main/kotlin/ru/testing/testlib/domain/User.kt [hunk 10]
```diff
@@ -1,9 +1,7 @@
 package ru.testing.testlib.domain
 
-// TODO: We need a domain module, probably.
-// Placing it into polygon will result in cyclic dependency between configuration and polygon.
 data class User(
-    val id: Int,
+    val id: Long,
     val name: String,
     val passwordHash: String
 )
```
Comment (zhelenskiy) [inline testlib/src/main/kotlin/ru/testing/testlib/domain/User.kt]: Empty line

File: testlib/src/main/kotlin/ru/testing/testlib/task/SubmissionVerdict.kt [hunk 11]
```diff
@@ -1,34 +1,39 @@
...
 
     /**
      * Not launched verdict
-     *
-     * @property testsCount count of tests
      */
-    data class NotLaunchedVerdict(private val testsCount: Int) : SubmissionVerdict()
+    @Serializable
+    object NotLaunchedVerdict : SubmissionVerdict()
 }
```
Comment (zhelenskiy) [inline testlib/src/main/kotlin/ru/testing/testlib/task/SubmissionVerdict.kt]: Empty line

**Review by zhelenskiy — CHANGES_REQUESTED** [review]

File: model/src/main/resources/database.properties (hunk 7 above)
Comment (vladrus13) [reply model/src/main/resources/database.properties:4]: OK

**Pull request merged**